File: backend/src/services/travelStatistics.ts
import type { EntryKind, Status, TripStage, TripTransport } from "@prisma/client";
import { parseInsights, type EntryInsights, type TransportKind } from "./redaction.js";

/**
 * Les statistiques du profil — la feuille « Statistiques », réservée aux
 * abonnés.
 *
 * **Rien n'est stocké : tout s'additionne à la lecture.** Chaque souvenir
 * rédigé porte son relevé (`Entry.insights`, écrit par l'agent de rédaction),
 * et cette passe les somme sur tous les voyages du compte. Un compteur tenu à
 * l'écriture aurait été plus rapide d'une milliseconde et faux à la première
 * suppression de souvenir ; ici, une correction ou une relance de la rédaction
 * se voit à la lecture suivante, sans rien à décompter.
 *
 * **Ce que le voyage déclare sert de plancher.** Un carnet dont aucun souvenir
 * n'a encore été rédigé montre déjà son pays et sa ville de destination, les
 * transports posés entre ses étapes et la distance de sa fiche : le relevé de
 * la rédaction **complète** ces déclarations, il ne les remplace pas. Une
 * carte de chiffres à zéro sur un voyage en cours se lirait comme une panne.
 */

export interface MemoForStatistics {
  id: string;
  stage: TripStage;
  startDate: Date | null;
  endDate: Date | null;
  dayCount: number | null;
  destinationCountryCode: string | null;
  destinationCity: string | null;
  distanceKilometres: number | null;
  steps: { destinationCountryCode: string | null; transport: TripTransport | null }[];
  entries: {
    kind: EntryKind;
    redactionStatus: Status;
    capturedAt: Date;
    insights: unknown;
  }[];
}

/** Les cinq chiffres qui se lisent au global comme sur le voyage en cours. */
export interface TravelFigures {
  countries: number;
  regions: number;
  cities: number;
  encounters: number;
  distanceKilometres: number;
}

export interface TransportUsage {
  kind: TransportKind;
  /** Le nombre de trajets, ou `null` quand aucun souvenir ne l'a compté. */
  count: number | null;
}

export interface CurrentTripStatistics {
  id: string;
  startDate: string | null;
  endDate: string | null;
  /** « Rome » — la dernière ville où un souvenir situe le voyageur. */
  currentPlace: string | null;
  /** Le nombre de jours du voyage, pour « 2 jours validés sur 21 ». */
  dayCount: number;
  /** Les journées qui ont au moins un souvenir rédigé. */
  validatedDays: number;
  figures: TravelFigures;
  /** Les vocaux enregistrés sur ce voyage. */
  recordings: number;
  transports: TransportUsage[];
}

export interface TravelStatistics {
  tripCount: number;
  overall: TravelFigures;
  currentTrip: CurrentTripStatistics | null;
  /**
   * Combien de souvenirs attendent encore leur relevé. Tant qu'il y en a,
   * l'app relit la feuille à intervalle court : c'est ce qui fait apparaître
   * un chiffre « en instantané » sans tenir de connexion ouverte.
   */
  pendingDetections: number;
  updatedAt: string;
}

/** Un ensemble insensible à la casse et aux espaces, qui garde la première graphie. */
class LabelSet {
  private readonly labels = new Map<string, string>();

  add(label: string | null | undefined): void {
    const trimmed = label?.trim();
    if (!trimmed) return;
    const key = trimmed.toLowerCase();
    if (!this.labels.has(key)) this.labels.set(key, trimmed);
  }

  addAll(labels: Iterable<string>): void {
    for (const label of labels) this.add(label);
  }

  get size(): number {
    return this.labels.size;
  }

  values(): string[] {
    return [...this.labels.values()];
  }
}

/** Ce qu'un voyage apporte, avant addition. */
interface MemoTally {
  countries: LabelSet;
  regions: LabelSet;
  cities: LabelSet;
  encounters: number;
  distanceKilometres: number;
  recordings: number;
  transports: Map<TransportKind, number | null>;
  currentPlace: string | null;
  validatedDays: number;
  pending: number;
}

const MILLISECONDS_PER_DAY = 86_400_000;

function dayKey(date: Date): string {
  return date.toISOString().slice(0, 10);
}

/**
 * Le nombre de jours d'un voyage : celui de sa fiche, sinon l'écart entre ses
 * deux dates, bornes comprises — un voyage du 10 au 12 dure trois jours.
 */
export function tripDayCount(memo: {
  dayCount: number | null;
  startDate: Date | null;
  endDate: Date | null;
}): number {
  if (memo.dayCount && memo.dayCount > 0) return memo.dayCount;
  if (!memo.startDate || !memo.endDate) return 0;
  const span = memo.endDate.getTime() - memo.startDate.getTime();
  return Math.max(0, Math.floor(span / MILLISECONDS_PER_DAY) + 1);
}

/**
 * Deux comptes de trajets s'additionnent ; un compte inconnu (`null`) ne
 * l'emporte jamais sur un compte connu, et deux inconnus restent inconnus —
 * « la semaine en scooter » deux fois reste « scooter », sans chiffre.
 */
function addTransport(
  tally: Map<TransportKind, number | null>,
  kind: TransportKind,
  count: number | null,
): void {
  const known = tally.get(kind);
  if (known === undefined) {
    tally.set(kind, count);
  } else if (count !== null) {
    tally.set(kind, (known ?? 0) + count);
  }
}

function tallyMemo(memo: MemoForStatistics): MemoTally {
  const tally: MemoTally = {
    countries: new LabelSet(),
    regions: new LabelSet(),
    cities: new LabelSet(),
    encounters: 0,
    distanceKilometres: 0,
    recordings: 0,
    transports: new Map(),
    currentPlace: null,
    validatedDays: 0,
    pending: 0,
  };

  // Le plancher : ce que le voyage déclare de lui-même.
  tally.countries.add(memo.destinationCountryCode?.toUpperCase());
  for (const step of memo.steps) tally.countries.add(step.destinationCountryCode?.toUpperCase());
  tally.cities.add(memo.destinationCity);

  const validatedDayKeys = new Set<string>();
  let latestPlaceAt = -Infinity;
  let detectedDistance = 0;
  let hasDetectedTransport = false;

  for (const entry of memo.entries) {
    if (entry.kind === "audio") tally.recordings += 1;
    if (entry.kind === "photo") continue;

    if (entry.redactionStatus === "pending" || entry.redactionStatus === "processing") {
      tally.pending += 1;
    }
    if (entry.redactionStatus === "ready") validatedDayKeys.add(dayKey(entry.capturedAt));

    if (entry.insights === null || entry.insights === undefined) continue;
    const insights: EntryInsights = parseInsights(entry.insights);

    tally.countries.addAll(insights.countries.map((country) => country.code));
    tally.regions.addAll(insights.regions);
    tally.cities.addAll(insights.cities);
    tally.encounters += insights.peopleMet;
    detectedDistance += insights.distanceKilometres ?? 0;

    for (const transport of insights.transports) {
      hasDetectedTransport = true;
      addTransport(tally.transports, transport.kind, transport.count);
    }

    // Le souvenir le plus récent qui situe le voyageur l'emporte : c'est là
    // qu'il est « actuellement ».
    const at = entry.capturedAt.getTime();
    if (insights.currentPlace && at > latestPlaceAt) {
      latestPlaceAt = at;
      tally.currentPlace = insights.currentPlace;
    }
  }

  // Les kilomètres relevés priment sur ceux de la fiche ; la fiche ne sert
  // que tant que rien n'a été relevé. Les deux ne s'additionnent pas — ce
  // serait compter deux fois le même trajet.
  tally.distanceKilometres =
    detectedDistance > 0 ? detectedDistance : (memo.distanceKilometres ?? 0);

  // Même règle pour les transports : ce que les souvenirs racontent, sinon ce
  // que les étapes déclarent.
  if (!hasDetectedTransport) {
    for (const step of memo.steps) {
      if (step.transport) addTransport(tally.transports, step.transport, 1);
    }
  }

  tally.currentPlace ??= memo.destinationCity;
  tally.validatedDays = validatedDayKeys.size;

  return tally;
}

/**
 * Le voyage « en cours » est le plus récemment commencé de ceux qui le sont :
 * la même règle que l'accueil et que la carte de chiffres du profil.
 */
function pickCurrent(memos: MemoForStatistics[]): MemoForStatistics | undefined {
  return memos
    .filter((memo) => memo.stage === "ongoing")
    .sort((a, b) => (b.startDate?.getTime() ?? 0) - (a.startDate?.getTime() ?? 0))[0];
}

function roundKilometres(value: number): number {
  return Math.round(value);
}

function serializeTransports(tally: Map<TransportKind, number | null>): TransportUsage[] {
  // Les plus fréquents d'abord ; les non comptés ferment la marche, comme
  // « scooter » à la fin de la ligne de la maquette.
  return [...tally.entries()]
    .map(([kind, count]) => ({ kind, count }))
    .sort((a, b) => (b.count ?? -1) - (a.count ?? -1));
}

/**
 * Additionne les voyages d'un compte. Pure : c'est ce qui la rend testable
 * sans base, et ce qui garantit que la route ne fait qu'une requête.
 */
export function aggregateTravelStatistics(
  memos: MemoForStatistics[],
  now: Date = new Date(),
): TravelStatistics {
  const tallies = new Map(memos.map((memo) => [memo.id, tallyMemo(memo)]));

  const countries = new LabelSet();
  const regions = new LabelSet();
  const cities = new LabelSet();
  let encounters = 0;
  let distance = 0;
  let pending = 0;

  for (const tally of tallies.values()) {
    countries.addAll(tally.countries.values());
    regions.addAll(tally.regions.values());
    cities.addAll(tally.cities.values());
    encounters += tally.encounters;
    distance += tally.distanceKilometres;
    pending += tally.pending;
  }

  const current = pickCurrent(memos);
  const currentTally = current ? tallies.get(current.id) : undefined;

  return {
    tripCount: memos.length,
    overall: {
      countries: countries.size,
      regions: regions.size,
      cities: cities.size,
      encounters,
      distanceKilometres: roundKilometres(distance),
    },
    currentTrip:
      current && currentTally
        ? {
            id: current.id,
            startDate: current.startDate?.toISOString() ?? null,
            endDate: current.endDate?.toISOString() ?? null,
            currentPlace: currentTally.currentPlace,
            dayCount: tripDayCount(current),
            validatedDays: currentTally.validatedDays,
            figures: {
              countries: currentTally.countries.size,
              regions: currentTally.regions.size,
              cities: currentTally.cities.size,
              encounters: currentTally.encounters,
              distanceKilometres: roundKilometres(currentTally.distanceKilometres),
            },
            recordings: currentTally.recordings,
            transports: serializeTransports(currentTally.transports),
          }
        : null,
    pendingDetections: pending,
    updatedAt: now.toISOString(),
  };
}

/**
 * Ce que la route lit, et rien de plus : les colonnes que l'addition regarde.
 * Les textes des souvenirs — transcription, rédaction, correction — n'en font
 * pas partie, et c'est ce qui garde la requête légère sur un compte qui a
 * raconté trois cents vocaux.
 */
export const memoStatisticsSelect = {
  id: true,
  stage: true,
  startDate: true,
  endDate: true,
  dayCount: true,
  destinationCountryCode: true,
  destinationCity: true,
  distanceKilometres: true,
  steps: { select: { destinationCountryCode: true, transport: true } },
  entries: {
    select: { kind: true, redactionStatus: true, capturedAt: true, insights: true },
  },
} as const;

File: ios/Modules/Sources/MemoBookCore/TravelStatistics.swift
import Foundation

// Ce que la feuille « Statistiques » du profil affiche, modélisé comme
// `GET /v1/profile/statistics` le rend.
//
// **Rien ici n'est calculé par l'app.** Chaque souvenir rédigé porte le relevé
// de l'agent de rédaction — pays, régions, villes, rencontres, kilomètres,
// transports, et l'endroit où le voyageur se trouve — et c'est le serveur qui
// les additionne à la lecture (`services/travelStatistics.ts`). L'app dessine
// ces chiffres, les accorde, et les relit tant que le serveur dit qu'un
// souvenir attend encore son relevé.

/// Les cinq chiffres qui se lisent au global comme sur le voyage en cours.
public struct TravelFigures: Codable, Sendable, Hashable {
    public var countries: Int
    public var regions: Int
    public var cities: Int
    /// Combien de personnes rencontrées — les compagnons de voyage n'y sont pas.
    public var encounters: Int
    /// Arrondis au kilomètre par le serveur ; l'app choisit seulement comment
    /// les écrire.
    public var distanceKilometres: Int

    public init(
        countries: Int = 0,
        regions: Int = 0,
        cities: Int = 0,
        encounters: Int = 0,
        distanceKilometres: Int = 0
    ) {
        self.countries = countries
        self.regions = regions
        self.cities = cities
        self.encounters = encounters
        self.distanceKilometres = distanceKilometres
    }

    public static let empty = TravelFigures()
}

/// Un moyen de transport employé sur le voyage en cours.
///
/// **Le nombre est facultatif, et c'est le sens de la ligne** : « 2 trains »
/// se compte, « scooter » se constate — la maquette écrit les deux sur la même
/// ligne. Le relevé ne pose un nombre que quand le récit l'a donné.
public struct TransportUsage: Codable, Sendable, Hashable, Identifiable {
    /// Les moyens que l'agent sait nommer. Fermée : c'est l'app qui en écrit
    /// le libellé et l'accord, et un moyen inconnu ne se lirait pas.
    public enum Kind: String, Codable, Sendable, Hashable, CaseIterable {
        case plane, train, bus, car, boat, bike, walk, scooter, motorbike, metro, taxi

        /// Le mot au singulier et son pluriel — « avion », « avions ». Les deux
        /// écrits, plutôt qu'un « s » collé : « bateau » fait « bateaux », et
        /// « à pied » ne se compte pas.
        public var words: (singular: String, plural: String) {
            switch self {
            case .plane: ("avion", "avions")
            case .train: ("train", "trains")
            case .bus: ("bus", "bus")
            case .car: ("voiture", "voitures")
            case .boat: ("bateau", "bateaux")
            case .bike: ("vélo", "vélos")
            case .walk: ("à pied", "à pied")
            case .scooter: ("scooter", "scooters")
            case .motorbike: ("moto", "motos")
            case .metro: ("métro", "métros")
            case .taxi: ("taxi", "taxis")
            }
        }
    }

    public let kind: Kind
    public var count: Int?

    public var id: Kind { kind }

    public init(kind: Kind, count: Int? = nil) {
        self.kind = kind
        self.count = count
    }

    /// « 1 avion », « 2 trains », « scooter ». Le nombre ne s'écrit que s'il a
    /// été relevé ; la marche ne se compte jamais.
    public var label: String {
        let (singular, plural) = kind.words
        guard let count, kind != .walk else { return singular }
        return count == 1 ? "\(count) \(singular)" : "\(count) \(plural)"
    }
}

/// Le voyage en cours, tel que la moitié basse de la feuille le montre.
public struct CurrentTripStatistics: Codable, Sendable, Hashable, Identifiable {
    public let id: String
    public var startDate: Date?
    public var endDate: Date?
    /// « Rome » — la dernière ville où un souvenir situe le voyageur.
    public var currentPlace: String?
    /// Le nombre de jours du voyage, pour « 2 jours validés sur 21 ».
    public var dayCount: Int
    /// Les journées qui ont au moins un souvenir rédigé.
    public var validatedDays: Int
    public var figures: TravelFigures
    /// Les vocaux enregistrés sur ce voyage.
    public var recordings: Int
    public var transports: [TransportUsage]

    public init(
        id: String,
        startDate: Date? = nil,
        endDate: Date? = nil,
        currentPlace: String? = nil,
        dayCount: Int = 0,
        validatedDays: Int = 0,
        figures: TravelFigures = .empty,
        recordings: Int = 0,
        transports: [TransportUsage] = []
    ) {
        self.id = id
        self.startDate = startDate
        self.endDate = endDate
        self.currentPlace = currentPlace
        self.dayCount = dayCount
        self.validatedDays = validatedDays
        self.figures = figures
        self.recordings = recordings
        self.transports = transports
    }

    /// Décodage tolérant : un transport que cette version de l'app ne connaît
    /// pas est **sauté**, pas fatal. L'agent peut apprendre un mot avant que
    /// l'app ne soit mise à jour, et une ligne de moins vaut mieux qu'une
    /// feuille qui ne s'ouvre plus.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        startDate = try container.decodeIfPresent(Date.self, forKey: .startDate)
        endDate = try container.decodeIfPresent(Date.self, forKey: .endDate)
        currentPlace = try container.decodeIfPresent(String.self, forKey: .currentPlace)
        dayCount = try container.decodeIfPresent(Int.self, forKey: .dayCount) ?? 0
        validatedDays = try container.decodeIfPresent(Int.self, forKey: .validatedDays) ?? 0
        figures = try container.decodeIfPresent(TravelFigures.self, forKey: .figures) ?? .empty
        recordings = try container.decodeIfPresent(Int.self, forKey: .recordings) ?? 0
        transports =
            try container.decodeIfPresent([Lenient<TransportUsage>].self, forKey: .transports)?
            .compactMap(\.value) ?? []
    }

    /// La part du voyage déjà racontée, entre 0 et 1 : les jours validés sur
    /// les jours du voyage. Zéro tant qu'on ne sait pas combien il dure — un
    /// pourcentage d'une durée inconnue ne veut rien dire.
    public var writtenFraction: Double {
        guard dayCount > 0 else { return 0 }
        return min(1, max(0, Double(validatedDays) / Double(dayCount)))
    }
}

/// Tout ce que la feuille « Statistiques » montre, d'un seul tenant.
public struct TravelStatistics: Codable, Sendable, Hashable {
    public var tripCount: Int
    /// Tous les voyages du compte, additionnés.
    public var overall: TravelFigures
    public var currentTrip: CurrentTripStatistics?
    /// Combien de souvenirs attendent encore leur relevé. Tant qu'il y en a,
    /// la feuille se relit à intervalle court : c'est ce qui fait bouger un
    /// chiffre sous les yeux, sans connexion ouverte avec le serveur.
    public var pendingDetections: Int
    public var updatedAt: Date?

    public init(
        tripCount: Int = 0,
        overall: TravelFigures = .empty,
        currentTrip: CurrentTripStatistics? = nil,
        pendingDetections: Int = 0,
        updatedAt: Date? = nil
    ) {
        self.tripCount = tripCount
        self.overall = overall
        self.currentTrip = currentTrip
        self.pendingDetections = pendingDetections
        self.updatedAt = updatedAt
    }

    /// Même raison que ``TravellerProfile`` : un serveur plus ancien, ou un
    /// champ ajouté plus tard, ne doit pas vider la feuille.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        tripCount = try container.decodeIfPresent(Int.self, forKey: .tripCount) ?? 0
        overall = try container.decodeIfPresent(TravelFigures.self, forKey: .overall) ?? .empty
        currentTrip = try container.decodeIfPresent(CurrentTripStatistics.self, forKey: .currentTrip)
        pendingDetections = try container.decodeIfPresent(Int.self, forKey: .pendingDetections) ?? 0
        updatedAt = try container.decodeIfPresent(Date.self, forKey: .updatedAt)
    }

    /// `true` tant que l'agent a encore des souvenirs à relire — la feuille
    /// le dit, et se relit.
    public var isDetecting: Bool { pendingDetections > 0 }

    /// Ce qui **compte** pour dire que la feuille a changé : les chiffres, pas
    /// l'horodatage. Deux lectures à trois secondes d'écart sur un voyage
    /// immobile ne doivent pas faire clignoter l'écran.
    public func hasSameFigures(as other: TravelStatistics) -> Bool {
        tripCount == other.tripCount
            && overall == other.overall
            && currentTrip == other.currentTrip
            && pendingDetections == other.pendingDetections
    }
}

/// Décode une valeur **ou rien**, au lieu de faire échouer le tableau qui la
/// porte. C'est ce qui permet à une liste de transports de survivre à un mot
/// que l'app ne connaît pas encore.
private struct Lenient<Value: Decodable>: Decodable {
    let value: Value?

    init(from decoder: any Decoder) throws {
        value = try? Value(from: decoder)
    }
}
